File: demo/android/app/src/main/kotlin/test/demo/sdkv2/ia/sdk/IaClientMethods.kt
package test.demo.sdkv2.ia.sdk

import android.content.Intent
import de.ihreapotheken.sdk.integrations.api.IaSdk
import de.ihreapotheken.sdk.ordering.OrderingModule
import de.ihreapotheken.sdk.otc.OtcModule
import de.ihreapotheken.sdk.pharmacy.PharmacyModule
import de.ihreapotheken.sdk.rx.RxModule
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter client service call handler.
 */
internal class IaClientMethods(
    private val bindings: IaClientBindings,
) {
    /**
     * Collection of available method invocation identifiers.
     */
    @Suppress("EnumEntryName")
    enum class FlutterCall {
        /**
         * Allocates the SDK runtime resources.
         */
        initIaSdk,

        /**
         * Places a new [android.app.Activity] object into the navigation stack.
         */
        launchRoute,

        /**
         * Forwards a collection of prescription objects with the ia.de checkout services.
         */
        transferPrescriptions,
    }

    fun callHandler(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            FlutterCall.initIaSdk.name -> {
                val args = call.arguments
                if (args !is Map<*, *>) {
                    result.error(
                        "ARG_ERROR",
                        "Arguments for initIaSdk must be of Dictionary type.",
                        null,
                    )
                    return
                }
                val accessKey = args["accessKey"] as? String
                val clientId = args["clientId"] as? String
                val serverEnvString = args["serverEnvironment"] as? String
                if (accessKey == null || clientId == null || serverEnvString == null) {
                    result.error(
                        "ARG_ERROR",
                        "Missing or invalid argument types. Expected String values for accessKey, clientId, and serverEnvironment.",
                        null,
                    )
                    return
                }
                bindings.sdkModule = IaSdk.register(
                    OtcModule,
                    OrderingModule,
                    PharmacyModule,
                    RxModule,
                )
                bindings.sdkModule.init(
                    context = bindings.applicationContext,
                    apiKey = accessKey,
                    clientID = clientId,
                )
                result.success(null)
            }

            FlutterCall.launchRoute.name -> {
                val viewId = call.arguments
                if (viewId !is String) {
                    result.error(
                        "ARG_ERROR",
                        "View identifier must be provided as a String type argument \"viewId\".",
                        null,
                    )
                    return
                }
                val intent = Intent(
                    bindings.applicationContext,
                    IaClientComponentActivity::class.java,
                )
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                intent.putExtra("viewId", viewId)
                bindings.applicationContext.startActivity(intent)
                result.success(null)
            }

            FlutterCall.transferPrescriptions.name -> {
                val data = call.arguments
                if (data !is Map<*, *>) {
                    result.error(
                        "ARG_ERROR",
                        "Prescription data must be provided as a Map<*, *> type argument.",
                        null,
                    )
                    return
                }
                val prescriptionImages = data["images"]
                if (prescriptionImages != null && (prescriptionImages !is ArrayList<*> ||
                            prescriptionImages.any { it !is ByteArray })
                ) {
                    result.error(
                        "ARG_ERROR",
                        "Prescription image data must be provided as a ArrayList<ByteData> type argument \"images\".",
                        null,
                    )
                }
                val prescriptionPdfs = data["pdfs"]
                if (prescriptionPdfs != null && (prescriptionPdfs !is ArrayList<*> ||
                            prescriptionPdfs.any { it !is ByteArray })
                ) {
                    result.error(
                        "ARG_ERROR",
                        "Prescription PDF data must be provided as a ArrayList<ByteData> type argument \"pdfs\".",
                        null,
                    )
                }
                val prescriptionCodes = data["codes"]
                if (prescriptionCodes != null && (prescriptionCodes !is ArrayList<*> ||
                            prescriptionCodes.any { it !is ArrayList<*> } ||
                            prescriptionCodes.all { (it as ArrayList<*>).all { it !is String } })
                ) {
                    result.error(
                        "ARG_ERROR",
                        "Prescription code data must be provided as a ArrayList<ArrayList<String>> type argument \"pdfs\".",
                        null,
                    )
                }
                @Suppress("UNCHECKED_CAST")
                bindings.sdkModule.transferPrescriptions(
                    images = prescriptionImages as ArrayList<ByteArray>,
                    pdfs = prescriptionPdfs as ArrayList<ByteArray>,
                    codes = prescriptionCodes as ArrayList<ArrayList<String>>
                )
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }
}

File: ios/appsdk_v2_flutter_plugin/Sources/app_sdk_v2_flutter_plugin/IaClientMethods.swift
import Combine
import Flutter
import IACore
import IAIntegrations
import IAOrdering
import IAOverTheCounter
import IAPharmacy

/// Flutter client service call handler.
@MainActor
internal class IaClientMethods {
    /**
     * Collection of available method invocation identifiers.
     */
    enum FlutterCall: CaseIterable {
        /**
         * Allocates the SDK runtime resources.
         */
        case initIaSdk

        /**
         * Registers SDK modules for use in the application.
         */
        case register

        /**
         * Selects a pharmacy by providing an identifier.
         */
        case setPharmacyId
        
        /**
         * Resets the state of user cart, clearing any added products or prescriptions.
         */
        case clearCart
        
        /**
         * Forwards the client personal information to the ia.de library for checkout purposes.
         */
        case setGuestUserData
        
        /**
         * Resets the user data and onboarding status (pharmacy selection, user consents statuses).
         */
        case logout
        
        /**
         * Places a new [UIViewController] object into the navigation stack.
         */
        case launchRoute
        
        /**
         * Forwards a collection of prescription objects with the ia.de checkout services.
         */
        case transferPrescriptions
        
        /**
         * Closes any overlaying ia.de screen contents.
         */
        case finishAllActivities

        /**
         * String identifier getter definition.
         */
        var name: String {
            return String(describing: self)
        }
    }
    
    /**
     * Flutter SDK host app bindings definitions.
     */
    private let bindings: IaClientBindings!
    private let argumentDecoder = IaArgumentDecoder()

    init(bindings: IaClientBindings!) {
        self.bindings = bindings
    }
    
    /**
     * Registers a handler for method calls from the Flutter side.
     */
    func callHandler(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            do {
                let returnValue = try await callHandlerInternal(call: call)
                result(returnValue)
            } catch {
                print(">>> callHandler error: \(error)")    // @TODO remove
                result(error.flutterError(methodName: call.method))
            }
        }
    }

    func callHandlerInternal(call: FlutterMethodCall) async throws -> Any? {
        switch call.method {
        case FlutterCall.register.name:
            let arguments = try argumentDecoder.decode(IaRegisterModulesArguments.self, from: call.arguments)
            let modules = arguments.modules
            
            // Map Flutter modules to SDK modules, doing it here and not in decoder because it requires all imports.
            var sdkModules = [IASDKModule]()
            for module in modules {
                switch module {
                case .integrations: sdkModules.append(.integrations)
                case .overTheCounter: sdkModules.append(.overTheCounter)
                case .ordering: sdkModules.append(.ordering)
                case .apofinder: sdkModules.append(.apofinder)
                case .pharmacyDetails: sdkModules.append(.pharmacyDetails)
                case .prescription: sdkModules.append(.prescription)
                case .cardLink: sdkModules.append(.cardLink)
                }
            }
            
            IASDK.register(sdkModules)
            return nil

        case FlutterCall.initIaSdk.name:
            let arguments = try argumentDecoder.decode(IaInitSdkArguments.self, from: call.arguments)
            let initializationOptions = arguments.initialization.mappedToSDK()
            
            // Set delegate
            IASDK.setDelegate(IaClientDelegate(channel: bindings.channel))
            
            // Configuration
            IASDK.setEnvironment(arguments.serverEnvironment.mappedToSDK())
            IASDK.configuration.apiKey = arguments.accessKey
            IASDK.configuration.clientID = arguments.clientId
            IASDK.configuration.shouldLoadRemoteStyleConfiguration = arguments.shouldFetchThemeFromRemote
            IASDK.configuration.footer.shouldShowAppSettings = arguments.footer.shouldShowAppSettings
            IASDK.configuration.footer.shouldShowDataProcessing = arguments.footer.shouldShowDataProcessing
            IASDK.configuration.footer.shouldShowImprint = arguments.footer.shouldShowImprint

            // @TODO: remove, just for testing
            IASDK.QA.setQAFeatures([.showTestPharmaciesOnApofinder])
            
            // Auto initialization is enabled to match Android behavior in terms of prerequisites.
            IASDK.configuration.isAutoInitializationEnabled = true
            IASDK.configuration.defaultInitializationOptions = initializationOptions

            // Initialization (we don't pass prerequisites options here to match android), isAutoInitializationEnabled is true which means it will run
            // prerequisites when some screen is shown.
            try await IASDK.initialize(shouldShowIndicator: initializationOptions.shouldShowIndicator, prerequisitesOptions: nil)
            return nil

        case FlutterCall.setPharmacyId.name:
            let arguments = try argumentDecoder.decode(IaSetPharmacyIdArguments.self, from: call.arguments)
            try await IASDK.Pharmacy.setPharmacyID(try arguments.pharmacyIdInt)
            return true

        case FlutterCall.clearCart.name:
            try await IAOrderingSDK.deleteCart()
            return nil

        case FlutterCall.setGuestUserData.name:
            let arguments = try argumentDecoder.decode(IaSetGuestUserDataArguments.self, from: call.arguments)
            try await IASDK.setUserData(arguments.mappedToSDK())
            return nil

        case FlutterCall.logout.name:
            try await IASDK.deleteAllUserRelatedData()
            return nil

        case FlutterCall.launchRoute.name:
            let arguments = try argumentDecoder.decode(IaLaunchRouteArguments.self, from: call.arguments)
            let screen = try arguments.view.iaScreen()
            screen.present()
            return nil

        case FlutterCall.transferPrescriptions.name:
            let arguments = try IaTransferPrescriptionsArguments(from: call.arguments)
            let mapped = arguments.mappedToSDK()

            try await IAOrderingSDK.transferPrescriptions(
                images: mapped.images,
                pdfs: mapped.pdfs,
                codes: mapped.codes,
                orderID: mapped.orderId,
                showActivityIndicator: true,
                finishAction: .openCart
            )
            return nil

        case FlutterCall.finishAllActivities.name:
            // @TODO: This will work for now but we will have to discuss how to best implement this.
            UIApplication.shared.rootViewController?.dismiss(animated: true)
            return nil

        default:
            return FlutterMethodNotImplemented
        }
    }
}
